### PR TITLE
DEX-570 Different fees for makers and takers

### PR DESCRIPTION
**Pull request by DenisSerebryansky**
Main:

  * Order fee settings now can be changed at some offset
  * Dynamic fee settings in application.conf now have 2 base fee instead of 1: base-maker-fee/base-taker-fee
  * Orders are validated by actual (current offset + 1) fee settings, fee is charged according to the current fee settings. Note, it can lead to matcher loses!
  * Method `getMakerTakerFee` in Matcher introduced. In case of dynamic fee settings matcher charges signed fee proportional to executed amount and _maker/taker ratio_
  * OrderBook accepts `getMakerTakerFee` lambda in `add` method and creates OrderExecuted event by fee settings associated with offset of the submitted order

Miscellaneous:

  * RateCacheSpecification introduced

### CODE REVIEW

File: dex-it/src/test/scala/com/wavesplatform/it/sync/orders/MakerTakerFeeTestSuite.scala [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  }
+
+  private val doubledFee: Long = 2 * matcherFee
+  private val maker            = bob
+  private val taker            = alice
+
+  private def test(makerAmount: Long, takerAmount: Long, takerFee: Long, isTakerMarket: Boolean = false): Unit = {
+
+    dex1.api.cancelAll(maker)
+    dex1.api.cancelAll(taker)
```
Comment (ngadiyak) [inline dex-it/src/test/scala/com/wavesplatform/it/sync/orders/MakerTakerFeeTestSuite.scala]: it seems that can be moved to afterEach method

File: dex/src/main/scala/com/wavesplatform/dex/Matcher.scala [hunk 2]
```diff
@@ -119,7 +120,7 @@ class Matcher(settings: MatcherSettings)(implicit val actorSystem: ActorSystem)
       assetPair,
       updateOrderBookCache(assetPair),
       marketStatuses.put(assetPair, _),
-      settings,
+      settings.orderFee match { case DynamicSettings(_, zeroMakerFee) => zeroMakerFee; case _ => false },
```
Comment (vsuharnikov) [inline dex/src/main/scala/com/wavesplatform/dex/Matcher.scala]: Minor: probably move this logic to a method `MatcherSettings.zeroMakerFee`

File: dex/src/test/scala/com/wavesplatform/dex/settings/BaseSettingsSpecification.scala [hunk 3]
```diff
@@ -18,6 +18,7 @@ class BaseSettingsSpecification extends FlatSpec {
        |  mode = percent
        |  dynamic {
        |    base-fee = 300000
+       |    zero-maker-double-taker = false
```
Comment (vsuharnikov) [inline dex/src/test/scala/com/wavesplatform/dex/settings/BaseSettingsSpecification.scala]: For tests it's better to specify a value different from application.conf's

File: dex/src/test/scala/com/wavesplatform/dex/MatcherSpecBase.scala [hunk 4]
```diff
@@ -48,8 +47,8 @@ trait MatcherSpecBase extends NTPTime with DiffMatcherWithImplicits with DoubleO
   protected val usd: IssuedAsset = mkAssetId("WUSD")
   protected val eth: IssuedAsset = mkAssetId("WETH")
 
-  protected val pairWavesBtc: AssetPair = AssetPair(Waves, btc)
-  protected val pairWavesUsd: AssetPair = AssetPair(Waves, usd)
```
Comment (vsuharnikov) [inline dex/src/test/scala/com/wavesplatform/dex/MatcherSpecBase.scala]: 👍

File: dex-it/src/test/scala/com/wavesplatform/it/sync/orders/MakerTakerFeeTestSuite.scala [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    val executedAmount = math.min(takerAmount, makerAmount)
+
+    withClue(s"maker amount = $makerAmount, taker${if (isTakerMarket) " (market)" else ""} amount = $takerAmount, taker fee = $takerFee\n") {
+      wavesNode1.api.balance(maker, Waves) shouldBe (makerInitialBalance - executedAmount)
+      wavesNode1.api.balance(taker, Waves) shouldBe (takerInitialBalance + executedAmount - takerFee)
+    }
+  }
+
+  "DEX should charge 0 fee for makers and doubled fee for takers" - {
+
```
Comment (ngadiyak) [inline dex-it/src/test/scala/com/wavesplatform/it/sync/orders/MakerTakerFeeTestSuite.scala]: uncovered cases:
 - market order
 - partial fill
 - possibility of being the taker with zero waves balance
 - fee in non-waves asset 
 - taker become a maker (is it possible?)

File: dex/src/test/scala/com/wavesplatform/dex/settings/BaseSettingsSpecification.scala (hunk 3 above)
Comment (DenisSerebryansky) [reply dex/src/test/scala/com/wavesplatform/dex/settings/BaseSettingsSpecification.scala]: Done

File: dex/src/main/scala/com/wavesplatform/dex/Matcher.scala (hunk 2 above)
Comment (DenisSerebryansky) [reply dex/src/main/scala/com/wavesplatform/dex/Matcher.scala]: Done

File: dex-it/src/test/scala/com/wavesplatform/it/sync/orders/MakerTakerFeeTestSuite.scala (hunk 1 above)
Comment (DenisSerebryansky) [reply dex-it/src/test/scala/com/wavesplatform/it/sync/orders/MakerTakerFeeTestSuite.scala]: Done

File: dex/src/main/scala/com/wavesplatform/dex/caches/RateCache.scala [hunk 6]
```diff
@@ -37,7 +38,7 @@ object RateCache {
   def apply(db: DB): RateCache = new RateCache {
 
     private val rateDB  = RateDB(db)
-    private val rateMap = new ConcurrentHashMap[IssuedAsset, Double](rateDB.getAllRates.asJava)
+    private val rateMap = new ConcurrentHashMap[IssuedAsset, Double] unsafeTap (_.putAll(rateDB.getAllRates.asJava))
```
Comment (vsuharnikov) [inline dex/src/main/scala/com/wavesplatform/dex/caches/RateCache.scala]: I've looked at `ConcurrentHashMap.scala` and what I see:
```java
    /**
     * Creates a new map with the same mappings as the given map.
     *
     * @param m the map
     */
    public ConcurrentHashMap(Map<? extends K, ? extends V> m) {
        this.sizeCtl = DEFAULT_CAPACITY;
        putAll(m);
    }
```

It looks like you're doing the same thing, but in a more complex way. Isn't it? :)

Comment (DenisSerebryansky) [reply dex/src/main/scala/com/wavesplatform/dex/caches/RateCache.scala]: Done

**Pull request merged**